### PR TITLE
Add pg_rewind to GPDB

### PR DESCRIPTION
**Pull request by l-wang**
Adding `pg_rewind`. It works for heap tables (ao/co tables are yet to be considered). Tests from the repository work and enabled to be run as part of ICW.

Check for target segment to be cleanly shut down has been disabled in gpdb. Since there is no plan to directly expose `pg_rewind` as a tool. But instead will be used by `gprecoverseg` for incremental recovery. `gprecoverseg` makes sure using catalog and also shuts down if target segment if its already running. The check to have it cleanly shutdown doesn't meet the needs of incrementally recovering the segment. We wish to recovery the segment even if it was not cleanly shutdown.

Work to use it as part of gprecoverseg will be done as separate PR.

We are currently adding it to contrib. 9.5+ in upstream added this in `src/bin`. There are pros and cons of adding having it in contrib or `src/bin`, we would like to know opinions what is preferred. 

Pros of having it in contrib:
- 9.4 where source is pulled from have it in contrib
- avoids merge complications for 9.5+

Cons of having it in contrib:
- Modifications made to pg_rewind code for GPDB in contrib will need to be manually carried someway to src/bin when 9.5 is merged
- contrib is considered optional think vs pg_rewind will always be needed for gpdb `gprecoverseg` tool

Also, we wish to know if any licensing additions needs to happen for these files.

### CODE REVIEW

File: contrib/pg_rewind/sql/basictest.sql [hunk 1]
```diff
@@ -42,17 +42,17 @@ function after_promotion
 # Insert a row in the old master. This causes the master and standby to have
 # "diverged", it's no longer possible to just apply the standy's logs over
 # master directory - you need to rewind.
-$MASTER_PSQL -c "INSERT INTO tbl1 VALUES ('in master, after promotion');"
+PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "INSERT INTO tbl1 VALUES ('in master, after promotion');"
 
 # Also insert a new row in the standby, which won't be present in the old
 # master.
-$STANDBY_PSQL -c "INSERT INTO tbl1 VALUES ('in standby, after promotion');"
+PGOPTIONS=${PGOPTIONS_UTILITY} $STANDBY_PSQL -c "INSERT INTO tbl1 VALUES ('in standby, after promotion');"
```
Comment (berlin-ab) [inline contrib/pg_rewind/sql/basictest.sql]: Connections using `$STANDBY_PSQL` after the standby has been promoted probably no longer need to be in utility mode.

File: contrib/pg_rewind/sql/config_test.sh [hunk 2]
```diff
@@ -69,5 +69,16 @@ PG_VERSION_NUM=90401
 PORT_MASTER=`expr $PG_VERSION_NUM % 16384 + 49152`
 PORT_STANDBY=`expr $PORT_MASTER + 1`
 
+PGOPTIONS_UTILITY='-c gp_session_role=utility'
 MASTER_PSQL="psql -a --no-psqlrc -p $PORT_MASTER"
 STANDBY_PSQL="psql -a --no-psqlrc -p $PORT_STANDBY"
+
+function wait_until_standby_is_promoted {
```
Comment (berlin-ab) [inline contrib/pg_rewind/sql/config_test.sh]: Very nice.

File: contrib/pg_rewind/sql/run_test.sh [hunk 3]
```diff
@@ -58,7 +51,7 @@ EOF
 echo "Master initialized."
 before_master
 
-pg_ctl -w -D $TEST_MASTER start >>$log_path 2>&1
+pg_ctl -w -D $TEST_MASTER start -o "-p ${PORT_MASTER} --gp_dbid=2 --gp_num_contents_in_cluster=3 --gp_contentid=0" >>$log_path 2>&1
```
Comment (berlin-ab) [inline contrib/pg_rewind/sql/run_test.sh]: The other magic numbers within these tests have been pulled out as variables into config_test.sh. We might consider doing the same here.

**Comment by Eulerizeit** [comment]
@pivotal-mike Alex has license questions. this is an FYI

File: contrib/pg_rewind/sql/basictest.sql (hunk 1 above)
Comment (l-wang) [reply contrib/pg_rewind/sql/basictest.sql]: The naming of master and standby in these tests are confusing under the Greenplum context, they really mean primary and mirror segments. We have to use utility mode when connecting to a segment.

**Comment by l-wang** [comment]
> AO/AOCO verification is simply a matter of adding AO/AOCO tables to the tests or is there a larger process you are imagining?

I haven't really played with AO/AOCO tables yet, but I've been only imagining simply adding AO/AOCO tables to the tests as the next step.

File: contrib/pg_rewind/sql/run_test.sh (hunk 3 above)
Comment (l-wang) [reply contrib/pg_rewind/sql/run_test.sh]: Good catch will do!

**Comment by hlinnaka** [comment]
> Adding pg_rewind. It works for heap tables (ao/co tables are yet to
> be considered). Tests from the repository work and enabled to be run
> as part of ICW.
Cool!

> Check for target segment to be cleanly shut down has been disabled in
> gpdb. Since there is no plan to directly expose pg_rewind as a tool.
> But instead will be used by gprecoverseg for incremental recovery.
> gprecoverseg makes sure using catalog and also shuts down if target
> segment if its already running. The check to have it cleanly shutdown
> doesn't meet the needs of incrementally recovering the segment. We
> wish to recovery the segment even if it was not cleanly shutdown.

I'm confused by this. It might be OK to remove the check that the target server is shut down cleanly, but no one's done a proper analysis to verify that in the community. Why take the risk here? If the target hasn't been cleanly shut down, gprecoverseg could start it up in single-user mode, to finish the crash recovery, and shut it down again.

Or if you do the analysis to make sure it's OK to remove that check, please submit a patch upstream to do the same there.

> We are currently adding it to contrib. 9.5+ in upstream added this in
> src/bin. There are pros and cons of adding having it in contrib or
> src/bin, we would like to know opinions what is preferred.
> 
> Pros of having it in contrib:
> 
> 9.4 where source is pulled from have it in contrib avoids merge
> complications for 9.5+
> 
> Cons of having it in contrib:
> 
> Modifications made to pg_rewind code for GPDB in contrib will need to
> be manually carried someway to src/bin when 9.5 is merged contrib is
> considered optional think vs pg_rewind will always be needed for gpdb
> gprecoverseg tool
> 
> Also, we wish to know if any licensing additions needs to happen for
> these files.

The reason you're asking about licensing is because this was copied from https://github.com/vmware/pg_rewind/tree/REL9_4_STABLE, not from PostgreSQL repository, right?. That's OK, it's licensed under the same PostgreSQL license as PostgreSQL itself. But please add a note somewhere in the comments to mention that.

But for technical reasons, did you consider backporting the PostgreSQL 9.5 version instead? I think the only part that you need from https://github.com/vmware/pg_rewind/tree/REL9_4_STABLE is the code in parsexlog.c, to decode each WAL record type. It might be easier going forward to use the PG 9.5 version. It's more or less the same, but there are a lot of small differences in function names, formatting, makefile, etc.

I would very much prefer to have this in src/bin/pg_rewing, like it is in the upstream. Having it in contrib avoids merge conflicts, but not really: we'll need to figure out which parts have been changed manually instead. We don't want to have two versions in the tree, after all.

**Comment by Eulerizeit** [comment]
@hlinnaka let me walk through the flow I believe you are suggesting.

We have a Primary/Mirror pair
Old Primary fails and the mirror is promoted to New Primary

We should:
1. restart the old primary letting it finish crash recovery
1. cleanly shut down the old primary
1. run pg_rewind unchanged. 

If this is the path followed upstream it certainly makes sense for us to follow that. Questions. 

If we restart in single user mode and allow the segment to finish crash recovery is there an easy way to check when it has finished?

**Comment by hlinnaka** [comment]
On 26 October 2018 15:05:26 WEST, Robert Eckhardt <notifications@github.com> wrote:
>@hlinnaka let me walk through the flow I believe you are suggesting.
>
>We have a Primary/Mirror pair
>Old Primary fails and the mirror is promoted to New Primary
>
>We should:
>1. restart the old primary letting it finish crash recovery
>1. cleanly shut down the old primary
>1. run pg_rewind unchanged. 
>
>If this is the path followed upstream it certainly makes sense for us
>to follow that. Questions. 

That's right.

>If we restart in single user mode and allow the segment to finish crash
>recovery is there an easy way to check when it has finished? 

Hmm. I guess if you launch the server like "postgres --single ...", and you just close its stdin without any commands, it will do crash recovery and exit with code 0 as soon that's done.


- Heikki

**Comment by ashwinstar** [comment]
On Fri, Oct 26, 2018 at 12:44 AM Heikki Linnakangas <
notifications@github.com> wrote:

> Adding pg_rewind. It works for heap tables (ao/co tables are yet to
> be considered). Tests from the repository work and enabled to be run
> as part of ICW.
> Cool!
>
> Check for target segment to be cleanly shut down has been disabled in
> gpdb. Since there is no plan to directly expose pg_rewind as a tool.
> But instead will be used by gprecoverseg for incremental recovery.
> gprecoverseg makes sure using catalog and also shuts down if target
> segment if its already running. The check to have it cleanly shutdown
> doesn't meet the needs of incrementally recovering the segment. We
> wish to recovery the segment even if it was not cleanly shutdown.
>
> I'm confused by this. It might be OK to remove the check that the target
> server is shut down cleanly, but no one's done a proper analysis to verify
> that in the community. Why take the risk here? If the target hasn't been
> cleanly shut down, gprecoverseg could start it up in single-user mode, to
> finish the crash recovery, and shut it down again.
>
For now, sure we can try to clean shut down. But seems overly strong
restriction. Given the pg_rewind workflow, we should try to lift it.
Understand its not done upstream yet.

I think this restriction exist because pg_rewind copies all the data pages
from current master since the two diverged and not really since last
checkpoint position before the divergence on standby. So, seems there is
room where some changes will be lost if not clean shutdown?

**Comment by ashwinstar** [comment]
On Fri, Oct 26, 2018 at 7:18 AM Heikki Linnakangas <notifications@github.com>
wrote:

> On 26 October 2018 15:05:26 WEST, Robert Eckhardt <
> notifications@github.com> wrote:
> >@hlinnaka let me walk through the flow I believe you are suggesting.
> >
> >We have a Primary/Mirror pair
> >Old Primary fails and the mirror is promoted to New Primary
> >
> >We should:
> >1. restart the old primary letting it finish crash recovery
> >1. cleanly shut down the old primary
> >1. run pg_rewind unchanged.
> >
> >If this is the path followed upstream it certainly makes sense for us
> >to follow that. Questions.
>
> That's right.
>
> >If we restart in single user mode and allow the segment to finish crash
> >recovery is there an easy way to check when it has finished?
>
> Hmm. I guess if you launch the server like "postgres --single ...", and
> you just close its stdin without any commands, it will do crash recovery
> and exit with code 0 as soon that's done.
>

Sure, will try that. If that's what's needed should be performed by
pg_rewind itself though instead of throwing the error. Wish to avoid steps
outside of it required by it. If works will propose upstream the same.

**Comment by l-wang** [comment]
Thanks all for review. We decide to use upstream 9.5 version of pg_rewind instead, so closing this PR in favor of https://github.com/greenplum-db/gpdb/pull/6107

**Pull request closed without merging**